internal/pkg/docker/orchestrator/orchestrator.go

```diff
@@ -452,6 +485,80 @@ func (o *Orchestrator) waitForContainerToStart(ctx context.Context, id string) e
 	}
 }
 
+func (o *Orchestrator) waitForContainerDependencies(ctx context.Context, containerName string, a *runTaskAction) error {
+	var deps []string
+	for depName, state := range a.task.Containers[containerName].DependsOn {
+		deps = append(deps, fmt.Sprintf("%s->%s", depName, state))
+	}
+	logMsg := strings.Join(deps, ", ")
+	fmt.Printf("Waiting for container %q dependencies: [%s]\n", containerName, color.Emphasize(logMsg))
+	eg, ctx := errgroup.WithContext(ctx)
+	for name, state := range a.task.Containers[containerName].DependsOn {
+		name, state := name, state
+		eg.Go(func() error {
+			ctrId := o.containerID(name)
+			isEssential := a.task.Containers[name].IsEssential
+			ticker := time.NewTicker(700 * time.Millisecond)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+				switch state {
+				case ctrStateStart:
+					err := o.waitForContainerToStart(ctx, ctrId, isEssential)
+					if err != nil {
+						return err
+					}
+					if isEssential {
+						log.Successf("Successfully container %s is running\n", ctrId)
+						return nil
+					}
+					fmt.Printf("non-essential container %q started successfully\n", ctrId)
+					return nil
+				case ctrStateHealthy:
+					healthy, err := o.docker.IsContainerHealthy(ctx, ctrId)
+					if err != nil {
+						if !isEssential {
+							fmt.Printf("non-essential container %q failed to be %q: %v\n", ctrId, state, err)
+							return nil
+						}
+						return fmt.Errorf("essential container %q failed to be %q: %w", ctrId, state, err)
+					}
+					if healthy {
+						log.Successf("Successfully dependency container %q reached %q\n", ctrId, state)
+						return nil
+					}
+				case ctrStateComplete:
+					exitCode, err := o.docker.IsContainerCompleteOrSuccess(ctx, ctrId)
+					if err != nil {
+						return fmt.Errorf("dependency container %q failed to be %q: %w", ctrId, state, err)
+					}
+					if exitCode != -1 {
```

Comment:
I see exitcode of `-1` means that the container is not running:
```go
// IsContainerCompleteOrSuccess returns true if a docker container exits with an exitcode.
func (c DockerCmdClient) IsContainerCompleteOrSuccess(ctx context.Context, containerName string) (int, error) {
	state, err := c.containerState(ctx, containerName)
	if err != nil {
		return 0, err
	}
	if state.Status == containerStatusRunning {
		return -1, nil
	}
	return state.ExitCode, nil
}
```
What about returning an error instead of `-1`? 

In addition, the function name `IsContainerCompleteOrSuccess` indicates that this is a boolean method; however in reality it returns an exit code. What do you think of refactoring it into:
```go
// ContainerExitCode returns the exit code of a container.
func (c DockerCmdClient) ContainerExitCode(ctx context.Context, name string) (int, error) {
	state, err := c.containerState(ctx, containerName)
	if err != nil {
		return 0, err
	}
	if state.Status == containerStatusRunning {
		return 0, errors.New("the container has not exited")
	}
	return state.ExitCode, nil
}
```